miner/scroll_worker.go

```diff
@@ -0,0 +1,943 @@
+	}
+	if w.prioritizedTx != nil {
+		from, _ := types.Sender(signer, w.prioritizedTx.tx) // error already checked before
+		txList := map[common.Address]types.Transactions{from: []*types.Transaction{w.prioritizedTx.tx}}
+		txs := types.NewTransactionsByPriceAndNonce(signer, txList, header.BaseFee)
+		if result := w.currentPipeline.TryPushTxns(txs, w.onTxFailingInPipeline); result != nil {
+			w.handleCCCResult(result)
+			return
+		}
+	}
+
+	if len(localTxs) > 0 {
+		txs := types.NewTransactionsByPriceAndNonce(signer, localTxs, header.BaseFee)
+		if result := w.currentPipeline.TryPushTxns(txs, w.onTxFailingInPipeline); result != nil {
+			w.handleCCCResult(result)
+			return
+		}
+	}
+	if len(remoteTxs) > 0 {
+		txs := types.NewTransactionsByPriceAndNonce(signer, remoteTxs, header.BaseFee)
+		if result := w.currentPipeline.TryPushTxns(txs, w.onTxFailingInPipeline); result != nil {
+			w.handleCCCResult(result)
+			return
+		}
+	}
+}
+
+func (w *worker) handleCCCResult(res *pipeline.Result) error {
+	if res != nil && res.OverflowingTx != nil {
+		if res.FinalBlock == nil {
+			// first txn overflowed the circuit, skip
+			log.Trace("Circuit capacity limit reached for a single tx", "tx", res.OverflowingTx.Hash().String(),
+				"isL1Message", res.OverflowingTx.IsL1MessageTx(), "reason", res.CCCErr.Error())
+
+			// Store skipped transaction in local db
+			overflowingTrace := res.OverflowingTrace
+			if !w.config.StoreSkippedTxTraces {
+				overflowingTrace = nil
+			}
+			rawdb.WriteSkippedTransaction(w.eth.ChainDb(), res.OverflowingTx, overflowingTrace, res.CCCErr.Error(),
+				w.currentPipeline.Header.Number.Uint64(), nil)
+
+			if overflowingL1MsgTx := res.OverflowingTx.AsL1MessageTx(); overflowingL1MsgTx != nil {
+				rawdb.WriteFirstQueueIndexNotInL2Block(w.eth.ChainDb(), w.currentPipeline.Header.ParentHash, overflowingL1MsgTx.QueueIndex+1)
+			} else {
+				w.eth.TxPool().RemoveTx(res.OverflowingTx.Hash(), true)
+			}
+		} else if !res.OverflowingTx.IsL1MessageTx() {
+			// prioritize overflowing L2 message as the first txn next block
+			// no need to prioritize L1 messages, they are fetched in order
+			// and processed first in every block anyways
+			w.prioritizedTx = &prioritizedTransaction{
+				blockNumber: w.currentPipeline.Header.Number.Uint64() + 1,
+				tx:          res.OverflowingTx,
+			}
+		}
+
+		switch {
+		case res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrBlockRowConsumptionOverflow):
+			l1TxRowConsumptionOverflowCounter.Inc(1)
+		case !res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrBlockRowConsumptionOverflow):
+			l2TxRowConsumptionOverflowCounter.Inc(1)
+		case res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrUnknown):
+			l1TxCccUnknownErrCounter.Inc(1)
+		case !res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrUnknown):
+			l2TxCccUnknownErrCounter.Inc(1)
```

Comment:
previously we `RemoveTx` for this case

Reply:
This switch-case is only for metrics collection.
Transaction is removed from the pool at [this](https://github.com/scroll-tech/go-ethereum/pull/735/files#diff-1825330f8f75827945cb0144a6ab3dc034a3081f48ee96b0a318c5de9950832eR789) line

Reply:
but the code you linked will only run when `res.FinalBlock == nil` (first txn overflowed the circuit)

what if the first several txns are okay, but then a following txn returns `circuitcapacitychecker.ErrUnknown`?

Reply:
All CCC errors are treated the same. If a txn fails CCC check, it gets processed as the first txn the next block. If it fails CCC check again as the first txn in the next block, if gets skipped and removed from the mempool. So, all txns get a second chance at going thru the CCC.